add: compose, readme, and dockerfile for qt gui

maffettone opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Adds the qtgui for the Queue Server from Bluesky widgets as an independent container.

## Description
<!--- Describe your changes in detail -->
Adds the image, docker compose, updated readme directions. Depends on [jozo/pyqt5](https://hub.docker.com/r/jozo/pyqt5). 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
People wanted better documentation around the gui, and having a quick container to spin up and test is the first step there. This should also facilitate better local testing. 

## Summary of Changes for Release Notes
<!--- Brief summary of changes that could be copied to the Release Notes. -->
<!--- Skip this section if this is a maintenace PR (CI, unit tests, typo fix etc.) -->
<!--- PRs with feature changes will not be merged without this section filled. -->
Add container for Bluesky Widgets QTGui (queue-monitor) for Bluesky Adaptive

<!--- Group the changes in the following sections: -->
### Added
- New docker image and entry point

### Changed
- docker compose file
- README

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested on a Mac which should be the hardest way to do this. 
- Follow readme. 
- Then docker compose up. 
- Use GUI to connect, then open environment. 
- Use GUI to add some agent driven naps to queue
- Start the queue

<!--
## Screenshots (if appropriate):
-->
![Screenshot 2023-12-11 at 10 36 53 AM](https://github.com/bluesky/bluesky-adaptive/assets/43007690/33f90b91-6db1-48b0-9246-1f48dc09f721)

Commit:
add: compose, readme, and dockerfile for qt gui

Commit:
maint: newline end

maffettone commented:
@padraic-shafer You can expect all tests to fail in the GHA. It is on the TODOs to fix up this test suite following the same paradigm as the containers used in local testing.

padraic-shafer commented on docker/README.md:
```diff
@@ -3,12 +3,24 @@
 The following docker files are split up for the queue-server and http-server for testing. The docker compose here will spin them up together along with a networked redis server. They are bare bones and will lack the full functionality of a collection profile, but should be useful for simple testing.
 
 ## To run the containers:
+On a Mac, to use the display you will need to install XQuartz and allow connections from network clients.
```
I think "Allow connections from network clients" might be a security vulnerability that will generally get flagged by cybersecurity scans.

maffettone replied:
This is on the money. I've fixed this to include the specific IP necessary prior to starting the container.

padraic-shafer replied:
Good idea!

padraic-shafer commented on docker/README.md:
```diff
@@ -3,12 +3,24 @@
 
 ## To run the containers:
+On a Mac, to use the display you will need to install XQuartz and allow connections from network clients.
+The queue-monitor gui image is built on [jozo/pyqt5](https://hub.docker.com/r/jozo/pyqt5). 
```
```suggestion
The queue-monitor gui container image is built on the base image [jozo/pyqt5](https://hub.docker.com/r/jozo/pyqt5). 
```

padraic-shafer commented on docker/README.md:
```diff
@@ -3,12 +3,24 @@
+docker build -t queue-monitor:latest .
 cd ../
+export LOCAL_DISPLAY_IP=$(ifconfig en0 | grep inet | awk '$1=="inet" {print $2}')
+xhost +
```
```suggestion
xhost + queue-monitor
```

For security I would add the client's hostname. I believe this should work with the name specified in the compose file.

maffettone replied:
Fixed.

padraic-shafer commented on docker/README.md:
```diff
@@ -3,12 +3,24 @@
+cd ../queue-monitor
+docker build -t queue-monitor:latest .
 cd ../
+export LOCAL_DISPLAY_IP=$(ifconfig en0 | grep inet | awk '$1=="inet" {print $2}')
```
Consider handing either Linux or MacOS here

```suggestion
export LOCAL_DISPLAY_IP=$([[ ! -z "${DISPLAY}" ]] && echo "${DISPLAY}" || ifconfig en0 | grep inet | awk '$1=="inet" {print $2":0"}')
```

maffettone replied:
This took a bit more wrangling because OSX has a `DISPLAY` variable that is not the IP needed for X11 forwarding from the container. So I broke this into a bash script that checks for Darwin, etc and operates accordingly.

padraic-shafer commented on docker/README.md:
````diff
@@ -3,12 +3,24 @@
+xhost +
 docker-compose up
 ```
+
+#### If using a Linux Machine, replace the `LOCAL_DISPLAY_IP` used for display:
+```bash
+export LOCAL_DISPLAY_IP=$DISPLAY
+```
````
Might no longer be necessary.  See comment above.

I believe the `:0` display ID is already included in `$DISPLAY`.

maffettone replied:
Sorted!

padraic-shafer commented on docker/docker-compose.yml:
```diff
@@ -24,6 +24,19 @@ services:
+    environment:
+      - QSERVER_ZMQ_CONTROL_ADDRESS=tcp://qserver:60615
+      - QSERVER_ZMQ_INFO_ADDRESS=tcp://qserver:60625
+      - DISPLAY=${LOCAL_DISPLAY_IP}:0
```
I believe the `:0` display ID is already included in `$DISPLAY`...meaning it would show up twice here.

See the suggestion in README.

padraic-shafer commented on docker/queue-monitor/entrypoint.sh:
```diff
@@ -0,0 +1,11 @@
+echo "Starting queue-monitor GUI..."
+echo "Control address: $QSERVER_ZMQ_CONTROL_ADDRESS"
+echo "Publish address: $QSERVER_ZMQ_INFO_ADDRESS"
+queue-monitor
```
I think will spawn `queue-monitor` each time a container command is run from the host.  Should this be a `CMD` instead?

maffettone replied:
Good thought, I've updated to use CMD.

padraic-shafer reviewed (commented):
I've posted a few questions and suggestions.

maffettone committed:
Update docker/README.md

Co-authored-by: Padraic Shafer <76011594+padraic-shafer@users.noreply.github.com>

Commit:
fix: display needs and security of xhost
- add: bash file to set local display conditionally on mac/linux
- Move entrypoint command to CMD
- Update README

padraic-shafer commented on docker/queue-monitor/local_display.sh:
```diff
@@ -0,0 +1,22 @@
+fi
+# If on a Mac, execute xhost + with LOCAL_DISPLAY_IP
+if [[ "$(uname)" == "Darwin" ]]; then
+    xhost + "$LOCAL_DISPLAY_IP"
```
I think the '+' and the host need to be contiguous here (but I'm not certain about that).

```suggestion
    xhost +"$LOCAL_DISPLAY_IP"
```

padraic-shafer reviewed (approved):
Thanks for making the changes.  The docs and container look really useful.

maffettone committed:
Update docker/queue-monitor/local_display.sh

Co-authored-by: Padraic Shafer <76011594+padraic-shafer@users.noreply.github.com>
